### PR TITLE
feat(color): Quick Model Select, keep active model in view when changing labels

### PR DESCRIPTION
**Pull request by philmoz**
On a number of occasions I have accidentally duplicated the current model from the popup menu.
This happens when I tap on the currently selected model and do not notice that the first entry is 'Duplicate Model'.

This change keeps the 'Select Model' entry as the first option even for the currently active model.
Selecting the menu option simply returns to the main view.

I think this makes for a more consistent UX.

### CODE REVIEW

**Comment by pfeerick** [comment]
I'm actually going to push a PR soon that will add a confirmation dialog to this (for the exact same reason), and a couple other more 'destructive' operations, so I'm not sure yet if this the right way to go.

**Comment by philmoz** [comment]
That's fine, it was just an idea.

Personally I think having consistent menus and behaviour is a better experience especially for new users.

**Comment by pfeerick** [comment]
That is true... but having menu entries that effectively do nothing can also be frustrating and clutter the UI... i.e. if there was a disabled state for menu entries that would probably be better. But then again, context-sensitive menus really should only offer up valid actions, rather than make the user have to guess which options will work or not work.

**Comment by ParkerEde** [comment]
@pfeerick @philmoz Please have a look to PR2626. That would solve this issue too. What do you think about it? Some of my EdgeTX User friends and I use this, and we are happy with it.

**Comment by philmoz** [comment]
@ParkerEde - thanks for the heads up. Personally I like your solution, I think for most people changing the active model is the primary use for the model select page, so that should be the easiest / quickest thing to do.

I like the idea of moving the 'delete model' button to the bottom of the list as well. It keeps the popup menu more consistent.

I had a play with your code and simplified it a bit. If you're ok with it I will add my version of your solution to this PR.

Having said all that, this does significantly change the way the model select page works which could be disruptive for users who are familiar with the current UX.

**Comment by philmoz** [comment]
Now with my adaptation of @ParkerEde ideas for model selection.

**Comment by ParkerEde** [comment]
I have tested this. It's great. But after Radio restart, the last Label is not selected(remembered) anymore.

What if we make a switch in radiosettings "Model quick select" False/True
True = short ENTER selects marked Model immediately, long ENTER shows Menu
False = short ENTER shows Menu
default is false.
So we would be maximally compatible and on the other hand it is possible to switch to the fast dialing behavior

**Comment by ParkerEde** [comment]
@philmoz 
After opening the model selection screen and selecting another model, the first ENTER keystroke is not executed. Only the second press of ENTER will select the model.

**Comment by ducatiJake** [comment]
Hi Guys, I made the code changes that @ParkerEde has commented on from PR #2626.  Basically pretty much the same functional changes, but your code changes are more straight forward than mine were, so I am happy with your changes.  I do agree with @ParkerEde on his above video that I also would prefer a single tap on a model icon box would select and open the tapped or clicked model.

**Comment by pfeerick** [comment]
Thanks for the video @ParkerEde 

So the only thing really missing here is some form of visual feedback as to that first button press/tap having been actioned? As I think there should be some visual/audible feedback to any input, especially one like this one... i.e. consider the situation where you don't realise the first press has happened (since there is no feedback of that having happened?) Perhaps a change in the border color? Is a beep/haptic if sound is enabled warranted?

**Comment by philmoz** [comment]
I've fixed the issue with the scroll wheel / enter button selection.
If you are just using the touch screen you still need to select the model twice - once to get focus, and then to select it as active. I found it was too sensitive otherwise, sometimes interpreting a scroll as a selection.

I also added a 'Model quick edit' setting as @ParkerEde suggested. Defaults to existing behaviour, turning it on selects new behaviour. Added this setting to companion as well.

Not sure I did the translation stuff correctly - I manually added the english string to all the language files.

**Comment by pfeerick** [comment]
At a glance, it looks fine... I don't think there is any need for the COLORLCD conditional - as there is no B&W alternative, but it won't do any harm. Very nice, thank you! I'll give it a try when I get back near a computer later.

**Comment by ParkerEde** [comment]
Translation DE
#define TR_MODEL_QUICK_SELECT        "schnelle Modellauswahl"

**Comment by ParkerEde** [comment]
Wow. 
I am very impressed. This looks super. 
Also tested it in Companion and Simulator for X10 and TX16S. Looks perfect to me. 
Thanks a lot
@pfeerick should we start a translator call?

**Comment by pfeerick** [comment]
@ParkerEde Yes, once there are no further changes needed to the PR. i.e. there is little point calling for translations if the initial string is wrong, or a second translation is needed ;) 

Is the behavioural change that "model quick select" off introduces intentional? As this is not consistent with the rest of the UI. For touch, first tap normally selects, second tap brings up the menu - here the first tap selects and brings up the menu. And for hard key navigation, first ENT to select, second ENT for menu. Check the file browser for comparison (I thought the theme manager was the same, but it seems not 🤷 - that probably needs tweaking). Quick mode touch and hard key navigation is working fine though. Companion read/write working on TX16S.

**Comment by pfeerick** [comment]
Thanks for that. One final request, and I think it's wrap. When selecting and unselecting a label, the active model doesn't always get brought into focus/shown... any chance of catching that or is it just an annoying edge case? Perhaps a video will explain it better... notice the model that has selection focus after the test label is unselected. 

https://drive.google.com/file/d/124jVpDwAq9mOSSLPajKvU2BA-IvopMuj/view?usp=sharing

**Comment by pfeerick** [comment]
Can we get some translations for this please 😃 

```
#define TR_MODEL_QUICK_SELECT "Model quick select"
```

@HThuren DA
~~@ParkerEde DE~~
@Eldenroot CZ
@ulfhedlund SE
@robustini IT
@zyren CN / TW
@ajjjjjjjj PL
@ToshihiroMakuuchi JP



image

image

**Comment by robustini** [comment]
IT
#define TR_MODEL_QUICK_SELECT "Selezione veloce modello"

**Comment by Eldenroot** [comment]
CZ: Rychlý výběr modelu

**Comment by ulfhedlund** [comment]
SE
#define TR_MODEL_QUICK_SELECT "Snabbval av modell"

**Comment by rotorman** [comment]
FI: "Mallin pikavalinta"

**Comment by zyren** [comment]
******************************************************************
CN
#define TR_MODEL_QUICK_SELECT "快速选择模型"
******************************************************************
******************************************************************
TW
#define TR_MODEL_QUICK_SELECT "快速選擇模型"
******************************************************************

**Comment by ToshihiroMakuuchi** [comment]
JP
#define TR_MODEL_QUICK_SELECT "モデル クイックセレクト"

**Comment by ajjjjjjjj** [comment]
PL
`#define TR_MODEL_QUICK_SELECT "Szybki wybór modelu"`

**Comment by HThuren** [comment]
DK
#define TR_MODEL_QUICK_SELECT "Hurtigvalg af model"

**Comment by philmoz** [comment]
@pfeerick The code tries to always have a model selected. On first load that will be the active model. If you change to a label that does not include the active model it selects the first model in the list. When you change back to other label, the selected model remains as the first model from the previous set, not the active model.

Thinking about it, this may not be the best UX. It might be better to always remember the last model selected by the user. If it is in the list when changing labels then it gets re-selected, otherwise no model will be selected until the user selects one.

**Comment by ducatiJake** [comment]
@philmoz That is the way my implementation worked also.  If the user has selected a model, then changed selected label, that included the currently selected model then that model stayed selected.  If the model was not in the new labels list, the first model in the list was selected.  Seemed right to me.  My code also tried to always have a selected model.  And by selected, I mean highlighted, not necessarily the active model.

**Comment by pfeerick** [comment]
> Thinking about it, this may not be the best UX. It might be better to always remember the last model selected by the user. If it is in the list when changing labels then it gets re-selected, otherwise no model will be selected until the user selects one.

No, as @ducatiJake suggested, I think this code actually behaves now is correct in that respect - there should always be a select model, especially for hard key navigation. I was just surprised it prioritised keeping the last selected model in view and selected, since it's not like you need it selected to drag it or do something else with it. More I'm trying to identify the practical use / benefit to the end user for that - it's not like you need to toggle into a different label to access some other context-sensitive menu or drag and drop or anything. Perhaps so that you can see if the model is in a different label or not (without actually checking the labels assigned to the model)? Otherwise, I would have thought the priority was would be the active model, if it's in the current label/filter.

**Comment by philmoz** [comment]
> Otherwise, I would have thought the priority was would be the active model, if it's in the current label/filter.

Either works for me. I've updated the PR to give priority to the active model.
Please try it and see which you prefer.

**Comment by ParkerEde** [comment]
There is still something not quite clean with the scroll direction of the marker. If I open the model selection and scroll backwards, the selection jumps from the first model to the "New" button (label). Once I have changed the label and am standing on the first model of the currently selected label and scroll backwards, the selection jumps to the "New model" button. I would say lastres should also happen after you just opened the model selection first.


https://user-images.githubusercontent.com/15068446/209087436-bf32d531-92d9-492c-a96f-61ef3187ba33.mp4

**Comment by philmoz** [comment]
The scrolling should be consistent now with the latest update to the PR. Please let me know how it goes.

**Comment by ParkerEde** [comment]
jepp, now it's consistent. The marker jumps always to the "New" (Label) Button. Tested with X10. 
I think it would be a little bit nicer, if the marker jumps always to the "New Model" Button instead. What do you think?

**Comment by philmoz** [comment]
> I think it would be a little bit nicer, if the marker jumps always to the "New Model" Button instead. What do you think?

The scroll order is based on the order objects are added to the display system. At the moment the ordering is top-bottom, left-right:
New Model <-> Labels <-> Sort Order <-> New Label <-> First Model ... Last Model <-> New Model

While the order can be changed it feels odd to me to have the scroll jump from First Model up to New Model.

**Comment by ParkerEde** [comment]
from the scroll order it would be strange indeed. However, I see it with the glasses of a user. Since the "New Model" button is placed above the models, I would think/expect the marker to jump there first. But that's not a showstopper for me now. So then we can leave it at that. Still very good job, thank you

**Review by pfeerick — APPROVED** [review]
Perfect timing - I was about to ask for a rebase ;) I agree re: "New Model" order... it for some reason would feel odd if it were changed from how it is... although I can also see the argument for it being changed. So let's leave it as it is for now until people start complaining it should be changed. 🤭 If everyone else is happy/thinks it makes sense that the focus should be the current model then this is ready to merge as far as I'm concerned.

**Comment by ducatiJake** [comment]
I built v2.8.1 this morning to check out these changes.  I seem to have the Select Model changes, but not the Model Quick Select system setting.  Not sure why.  Anyhow, I have noticed one inconsistency.  If I open Model Select, then use the roller to highlight another model, I have to click the roller twice to make it the active model and close the MS screen.  If I open Model Select then highlight another model by tapping on that model which causes the same visual change as if I had used the roller to highlight it I only have to tap the model or press the roller 1 more time to make it the active model and close the MS screen.  Shouldn't using the roller to highlight a model be the same as tapping on a model.  Both give the same highlighted visual feedback.  Seems to me that if a model is highlighted no matter how I got there a single tap or roller click should make that model active and close the MS screen.  Other than this, I love these changes.  I don't think the system setting is needed myself, but if most people believe it should be there, I have no issues with it being there.  If it defaults off, it just makes it harder for users to discover these improvements.

**Comment by ParkerEde** [comment]
@ducatiJake 
I suspect there is something wrong with the build with you. It is also included in Companion. However, I have only tested a Windows build. A model selected with the rotary encoder is opened with the first Enter keystroke. In case of touch operation, a model should be selected at the first touch and opened at the second touch. This is how it was agreed here. This is also how the current build behaves.
[libedgetx-tx16s-simulator.zip](https://github.com/EdgeTX/edgetx/files/10288447/libedgetx-tx16s-simulator.zip)
This libedgetx-tx16s-simulator.dll is based on actual main+this pr

**Comment by ducatiJake** [comment]
Thanks @ParkerEde I tried your .dll and see the same behavior as you.  My build just got main plus everything labeled 2.8.1 which at least when I did it was not complete.  Thanks for clarifying that.

**Comment by pfeerick** [comment]
This PR branch hasn't been merged yet, so unless you checked out this PR, you must have built against the wrong branch. Milestone is only a tracker of where it is planned this PR will first be in a release.

**Comment by HThuren** [comment]
hi @ParkerEde / @pfeerick @zyren 
Notice, this define are missing in tw.h
#define TR_MODEL_QUICK_SELECT "快速選擇模型"

**Comment by pfeerick** [comment]
Thanks @HThuren ... fixed in 4b838c73541282ba534a469f6d110b55aa74d2e5

File: radio/src/gui/colorlcd/model_select.cpp [hunk 1]
```diff
@@ -366,41 +367,46 @@ ModelsPageBody::ModelsPageBody(Window *parent, const rect_t &rect) :
...
-          }
-          return !TELEMETRY_STREAMING() || g_eeGeneral.disableRssiPoweroffAlarm;
-        })) {
-      return;  // stop if connected but not confirmed
+  // Don't need to check connection to receiver if re-selecting the active model
+  if (model != modelslist.getCurrentModel()) {
+    bool modelConnected =
+        TELEMETRY_STREAMING() && !g_eeGeneral.disableRssiPoweroffAlarm;
+    if (modelConnected) {
+      AUDIO_ERROR_MESSAGE(AU_MODEL_STILL_POWERED);
```
Comment (pfeerick) [inline radio/src/gui/colorlcd/model_select.cpp:379]: @philmoz I finally found a bug from this PR, which is quite nasty as it can completely lock the radio up. With the Quick Model select option on (it only triggers in the ON state), and you have a telemetry receiver active, when trying to switch model, you will get this warning, but it will keep repeating the message, and the haptic vibration continuously rather than a single time, and the radio will need to be hard powered off even if you do power off the receiver (and the prompts stop). This was with 2.8.1 - `main` does not seem to do this.

Comment (philmoz) [reply radio/src/gui/colorlcd/model_select.cpp:379]: I think that was fixed in PR #3061 with the changes to the full screen dialog.
One of the commits in that PR mentions fixing the infinite loop issue.

Comment (pfeerick) [reply radio/src/gui/colorlcd/model_select.cpp:379]: Thanks - I'll see if https://github.com/EdgeTX/edgetx/pull/3061/commits/96cfe35071e8e365d767f1daf5e20dc6eadef409 fixes it then and push it in for 2.8.2 if it behaves. :)

Comment (pfeerick) [reply radio/src/gui/colorlcd/model_select.cpp:379]: Now that you mention that PR, that reminds me... would you be able to double check if my memory is correct/why some of the alerts from  `storage/sdcard_yaml.cpp` : `const char * loadRadioSettingsYaml(bool checks)` don't have any touch buttons? I think it's the ones that come up if there aren't any radio settings present.

Comment (philmoz) [reply radio/src/gui/colorlcd/model_select.cpp:379]: There are three variation for the full screen popup:
- confirmation dialog (with Enter and RTN buttons), generated with confirmationDialog() call
- alert dialog with one action button defined by caller, generated with raiseAlert() and non-empty 3rd param
- alert dialog with no buttons, generated by ALERT macro or raiseAlert() with empty 3rd param

sdcard_yaml.cpp uses the ALERT macro.
